File: 659. Encode and Decode Strings/Main.java
import java.util.ArrayList;
import java.util.List;

class Main {
    public static void main(String[] args) {
        List<String> input = List.of("32#howdy", "wor%45#world");
        Solution solution = new Solution();
    
        String encoded = solution.encode(input);
        List<String> output = solution.decode(encoded);
        
        boolean status = true;
        if (input.size() == output.size()) {
            for (int i = 0; i < input.size(); i++) {
                status = status && (input.get(i).equals(output.get(i)));
            }
        } else {
            status = false;
        }
        System.out.println(status);
        System.out.println(">>>> input = " + input);
        System.out.println(">>>> output = " + output);
    }
}

class Solution {
    public String encode(List<String> strs) {
        StringBuilder builder = new StringBuilder();
        for (String str : strs) {
            builder.append(str.length()).append("#").append(str);
        }
        return builder.toString();
    }

    public List<String> decode(String str) {
        List<String> result = new ArrayList<>();
        int i = 0;

        while (i < str.length()) {
            int j = i;
            while (str.charAt(j) != '#') {
                j++;
            }

            int length = Integer.parseInt(str.substring(i, j));
            i = j + 1 + length;
            result.add(str.substring(j + 1, i));
        }

        return result;
    }
}

File: 659. Encode and Decode Strings/main.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func main() {
	input := []string{"32#howdy", "wor%45#world"}

	encoded := encode(input)
	output := decode(encoded)

	status := true
	if len(input) == len(output) {
		for i := 0; i < len(input); i++ {
			status = status && (input[i] == output[i])
		}
	} else {
		status = false
	}

	fmt.Println(status)
	fmt.Println(input)
	fmt.Println(output)
}

func encode(strs []string) string {
	builder := strings.Builder{}
	defer builder.Reset()

	for _, str := range strs {
		builder.WriteString(strconv.Itoa(len(str)))
		builder.WriteRune('#')
		builder.WriteString(str)
	}
	return builder.String()
}

func decode(str string) []string {
	var output []string
	i := 0

	for i < len(str) {
		j := i
		for str[j] != '#' {
			j++
		}

		length, _ := strconv.Atoi(str[i:j])
		i = j + 1 + length
		output = append(output, str[j+1:i])
	}
	return output
}
